lib/srv/sess.go

```diff
@@ -730,52 +536,89 @@ func (s *session) Recorder() events.StreamWriter {
 	return s.recorder
 }
 
-// Close ends the active session forcing all clients to disconnect and freeing all resources
-func (s *session) Close() error {
-	s.closeOnce.Do(func() {
-		serverSessions.Dec()
-		// closing needs to happen asynchronously because the last client
-		// (session writer) will try to close this session, causing a deadlock
-		// because of closeOnce
-		go func() {
-			s.mu.Lock()
-			defer s.mu.Unlock()
-
-			close(s.closeC)
-			s.BroadcastMessage("Closing session...")
-			s.log.Infof("Closing session %v.", s.id)
-			if s.term != nil {
-				s.term.Close()
-			}
+// Stop ends the active session and forces all clients to disconnect.
+// This will trigger background goroutines to complete session cleanup.
+func (s *session) Stop() {
+	s.mu.Lock()
+	defer s.mu.Unlock()
 
-			for _, p := range s.parties {
-				err := p.Close()
-				if err != nil {
-					s.log.WithError(err).Errorf("Failed to close party.")
-				}
-			}
+	select {
+	case <-s.stopC:
+		return
+	default:
+		close(s.stopC)
+	}
 
-			if s.recorder != nil {
-				s.recorder.Close(s.serverCtx)
-			}
-			s.state = types.SessionState_SessionStateTerminated
-			s.stateUpdate.Broadcast()
-			err := s.trackerUpdateState(types.SessionState_SessionStateRunning)
-			if err != nil {
-				s.log.Warnf("Failed to set tracker state to %v", types.SessionState_SessionStateRunning)
-			}
-		}()
-	})
-	return nil
+	s.BroadcastMessage("Stopping session...")
+	s.log.Infof("Stopping session %v.", s.id)
+
+	// close io copy loops
+	if s.io != nil {
```

Comment:
Can stop be called where io is nil at any point? It seems to be set at session creation and never to nil